chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,317 @@
+// Package chanfitness monitors the behaviour of channels to provide scores
+// which can be factored into the decision to close a channel. This is
+// achieved by maintaining an event store which tracks events for each
+// channel.
+//
+// Lifespan: the period that the channel has been known to the scoring system.
+// Note that lifespan may not equal the channel's full lifetime because data is
+// not currently persisted.
+//
+// Uptime: the total time within a given period that the channel's remote peer
+// has been online.
+package chanfitness
+
+import (
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/channelnotifier"
+	"github.com/lightningnetwork/lnd/peernotifier"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// ChannelEventStore maintains a set of event logs for the node's channels.
+// It is intended to provide insight into the performance of channels.
+type ChannelEventStore struct {
+	cfg *Config
+
+	// channels maps short channel IDs to event logs.
+	channels map[uint64]*chanEventLog
+
+	// peers tracks the current online status of peers based on online/offline
+	// events.
+	peers map[[33]byte]bool
+
+	quit chan struct{}
+
+	sync.Mutex
+
+	wg sync.WaitGroup
+}
+
+// Config provides the event store with functions required to monitor channel
+// activity. All elements of the config must be non-nil for the event store to
+// operate.
+type Config struct {
+	// Channel events provides a subscription client which provides a
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
+
+	// PeerEvents provides a subscription client which provides a stream of
+	// peer online/offline events.
+	PeerEvents func() (*subscribe.Client, error)
+
+	// Get open channels provides a list of existing open channels
+	// which is used to populate the ChannelEventStore with a set of channels
+	// on startup.
+	GetOpenChannels func() ([]*channeldb.OpenChannel, error)
```

Comment:
So get rid of having to pass a list of open channels to the store at all?
Or find a way to pipe them through the same logic without necessarily having to trigger an event somehow?

Reply:
It could be done, but that would break the 1:1 relationship to `SubscribeChanEvents` from `channelnotifier`, so I think the current way it is done is alright.

Reply:
Discussed offline. Conclusion is that the `ChannelNotifier` interface isn't optimal. You can subscribe to events, but there is no way to pass in a reference point from where you want to get those events. For example, in `RegisterBlockEpoch`, you can pass in a best block. This makes it clear what to expect. If `ChannelNotifier` would have had such a point (for example blockheight from which you receive channel events), the channel fitness subsystem could just pass in height 0 and get an initial sync via the same subscription client.

You could consider adding a TODO to `ChannelNotifier.SubscribeChannelEvents` to document the shortcoming. At least to prevent future devs from, without being fully aware, adding another `ListChannels` and event de-duplication in the next subsystem that uses channel notifier.

Reply:
+1 to keeping the current design with an additional TODO 👍

Reply:
TODO added to `channotifier`.